nodejs: fix cross compilation for armv7l

Cynerd opened the pull request:
###### Description of changes
The cross build for armv7l requires build to be on i686 (32bit in general). This ensures that and thus fixes build for armv7l.

I am not happy with this solution but I am unable to figure out how to fix cross compilation for amv7l in better way than this.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

alyssais reviewed (changes requested):
If we truly need to force a 32-bit build system, I think a cleaner way to do it would be to do it in the `callPackage` invocation.  Then we could just do something like

```
nodejs = (if stdenv.buildPlatform.isx86_64 && stdenv.is32bit then pkgsi686Linux else pkgs).callPackage …
```

which is nice because then there's one less package set involved in the actual expression, and it prevents adding 64-bit dependencies as well by mistake.

Either way, this needs a comment explaining what's going on, with a link to documentation or reference to the source code so people can understand what's going on, and where to check to see if it can be removed.

alyssais commented on pkgs/development/web/nodejs/nodejs.nix:
```diff
@@ -39,6 +39,10 @@ let
 
+  buildPkgs = if stdenv.buildPlatform.isx86_64 && stdenv.is32bit
+    then buildPackages.pkgsi686Linux.buildPackages
+    else buildPackages;
```
(Small note, that will be irrelevant if we use the callPackage version instead.) This is incorrect.  buildPackages is equivalent to nativeBuildInputs.  For depsBuildBuild, packages are sourced from pkgsBuildBuild.

Cynerd replied:
Call package version used instead.

Cynerd committed:
nodejs: fix cross compilation for armv7l

The v8 library supports only limited number of build platforms based on
the host architecture. The rule there is the bitness (the mixing of
32bit and 64bit architectures seems to be in most cases disallowed).
Thus this adds usage of the emulator of the host platform and builds
tools for that.

Cynerd commented:
> If we truly need to force a 32-bit build system, I think a cleaner way to do it would be to do it in the callPackage invocation. Then we could just do something like

I tweaked it a bit. It is probably reasonable requirement.

> Either way, this needs a comment explaining what's going on, with a link to documentation or reference to the source code so people can understand what's going on, and where to check to see if it can be removed.

I added comment but I am unable to find that this is actually required in the documentation. At the same time build really fails with missing 32bit stubs and thus it for sure tries to use 32bit build somewhere inside. The upstream builds it on Ubuntu and thus they might expect that 32bit libraries are just available while that is not the case with nixpkgs.

alyssais reviewed (approved):
LGTM but I'd really like there to be an upstream issue about this.  They shouldn't be passing -m32 to the native compiler just because they're cross compiling for 32 bit.

lilyinstarlight commented:
I only just now saw this, but I do wonder if it may also be fixed by #238377?

Cynerd commented:
@lilyinstarlight  I just tested it, and it doesn't help my issue. The issue here is with cross-compilation from amd64 to 32-bit systems. Node requires bitness to match it seems, and for that reason, we need to build it not in an amd64 environment but instead in an x86 one.

I suspect that, especially on new Macs that do not have 32-bit arm support (if I remember correctly) it won't even be possible to cross-compile Node to armv7l.

Cynerd commented:
> LGTM but I'd really like there to be an upstream issue about this. They shouldn't be passing -m32 to the native compiler just because they're cross compiling for 32 bit.

I create an issue upstream but I am not expecting much https://github.com/nodejs/node/issues/48612. This is a common issue I saw with other bytecode compilers (for example luajit).

tie commented:
FWIW I think this is the relevant code that sets -m32 when cross-compiling in your case.
https://github.com/nodejs/node/blob/3205b1936a5c3c21b0c8c568a57b0d2bada366b1/tools/v8_gypfiles/toolchain.gypi#L583-L601

host_cxx_is_biarch is defined at https://github.com/nodejs/node/blob/3205b1936a5c3c21b0c8c568a57b0d2bada366b1/tools/v8_gypfiles/toolchain.gypi#L107-L118

That is, it assumes that x86(-64) (and some other systems) have bi-arch capable toolchains (probably for a reason, see below, but maybe not, this is technical decision is really weird).

With the following patch, we successfully get a meaningful error (instead of vague compilation failure due to the missing headers) that leads us to https://github.com/nodejs/node/blob/3205b1936a5c3c21b0c8c568a57b0d2bada366b1/deps/v8/include/v8config.h#L863-L892 (or `deps/v8/src/base/build_config.h` for v18)

<details>

```patch
diff --git a/tools/v8_gypfiles/toolchain.gypi b/tools/v8_gypfiles/toolchain.gypi
index 26c6866fa9..658ebd6e86 100644
--- a/tools/v8_gypfiles/toolchain.gypi
+++ b/tools/v8_gypfiles/toolchain.gypi
@@ -109,7 +109,7 @@
       host_arch=="s390x" or \
       clang==1', {
       'variables': {
-        'host_cxx_is_biarch%': 1,
+        'host_cxx_is_biarch%': 0,
        },
      }, {
       'variables': {
```
</details>

> *Target architecture arm is only supported on arm and ia32 host*

So this seems like an issue with v8 rather than Node.js. At this point I don’t really want to dive deeper into this rabbit hole 🥲

tie commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9404,6 +9404,15 @@ with pkgs;
 
+  # The NodeJS packages has to be build in 32bit environment if host platform is
+  # also 32bit because it uses 32bit stubs and links against 32bit OpenSSL. The
+  # only architecture that supports execution of 32bit is x86_64 and thus that
```
```suggestion
  # only architecture that generally supports execution of 32bit is x86_64 and thus that
```

IIRC AArch64 can execute AArch32 code if enabled in the kernel configuration ([Kernel support for 32-bit EL0 found in arch/arm64/Kconfig](https://cateee.net/lkddb/web-lkddb/COMPAT.html))? That seems to be CPU-dependent for ARM, though I think 32-bit can also be disabled on x86-64.

Cynerd replied:
Yes, that is CPU-dependent on aarch64 because it relies on the CPU to actually interpret those arm instructions, as far as I know. The only way I know is to use binfmt and register an emulator. On the second note about disable of x86: you can even theoretically disable the https://www.kernelconfig.io/CONFIG_COMPAT_BINFMT_ELF?kernelversion=6.3.12 it just depends on where you draw the line between "this is expected platform support and this is not". Linux kernel provides way too many compilations knobs to cover all corners.

But I agree with the comment change (and it is done by now). This answer is only just a polemic.

lilyinstarlight commented:
@Cynerd Could you rebase and resolve merge conflicts on this?

If this is what we have to do, I'm mostly alright with this, but it is rather unfortunate

Cynerd commented:
@lilyinstarlight rebase is done.

tie commented on pkgs/top-level/all-packages.nix:
```diff
@@ -10007,20 +10007,30 @@ with pkgs;
+  # also 32bit because it uses 32bit stubs and links against 32bit OpenSSL. The
+  # only architecture that generally supports execution of 32bit is x86_64 and
+  # thus that is the only one handled here.
+  callPackageNodejs =
```
This is a public attribute that people may use even though it’s an internal implementation detail. I think it makes sense to move it to the `v{18,20,22,…}.nix` files (or at least avoid exposing it).

Cynerd replied:
That was a previous implementation that I changed on bases of these arguments https://github.com/NixOS/nixpkgs/pull/220204#pullrequestreview-1347419337.

tie replied:
>the standard callPackage is publicly available
Yes, because it is a generic public interface while `callPackageNodejs` is an alias used only to build Node.js package. Is there some other use case I’m missing here?

Cynerd replied:
Like packaging a custom version in overlays?

tie replied:
Isn’t `override{,Attrs}` enough in this case?

Cynerd replied:
You are right.

I will move it to the vxx.nix files.

tie commented:
Hm, I’ve just noticed how buildroot handles that, would it make sense to use `stdenv.hostPlatform.emulator` here? Is that even possible?

https://github.com/buildroot/buildroot/blob/ecf060a2ff93626fb5fbe1c52b8d42b6ee0fbb5b/package/nodejs/nodejs-src/nodejs-src.mk#L175-L204

tie commented:
>Although it might be required in the end considering that node closed the issue about support for cross compilation on different bitness and the need to also build it on not only x86 (amd64) architecture.

The change in this PR kinda overrides localSystem arguments of the Nixpkgs when building Node.js (see pkgsi686Linux definition in pkgs/top-level/stage.nix) and I’m not really comfortable with that, also https://github.com/NixOS/nixpkgs/pull/220204#discussion_r1256713670

>I am not sure how to transfer that directly to the nixpkgs because I would probably need to switch build from cross-compilation, but that would use all tools from the host platform.

Hm, at least getting the QEMU wrapper patch seems doable unless I’m missing something here.

- `v8-qemu-wrapper` translates directly to `stdenv.hostPlatform.emulator buildPackages`,
- `{HOST_,}NODEJS_SRC_WRAPPER_FIXUP` goes to the patchPhase (along with the accompanying `0001-add-qemu-wrapper-support.patch` in that directory).
- I don’t see any other unusual configure flags in buildroot package aside from them always using `--cross-compiling` and related flags (on that note, see also #317109 to avoid unnecessary rebases if someone with commit permissions finally sees it).

Cynerd commented:
> Hm, at least getting the QEMU wrapper patch seems doable unless I’m missing something here.

Yes, that is the easy part. The nasty part is in providing the host toolchain as a build one without actually using `callPackage` from the host platform (that would pull all dependencies and thus be effectively built in the qemu, which is really slow).

tie commented:
I don’t see how that would be slow given that only a handful of build steps actually need an emulator.

Cynerd commented:
> I don’t see how that would be slow given that only a handful of build steps actually need an emulator.

You misunderstood me. I was attempting to do that, and I was failing because the build system was mixing the host and build architectures.. at least, that is what I thought was happening. In the end, it turned out that I had an issue with the modification script, and it all went crashing down way later with some dependencies and errors about invalid architecture. My message was about, at that moment, seeing the only option to build it all in the host platform and thus emulate everything (that would be slow). In the end, I was able to solve it. I will push here the reworked version that uses an emulator to compile it. It started working once I fixed some stupid errors (missing commas).

tie commented:
Yes, should be targeting staging given the number of rebuilds.

See https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging

Cynerd commented:
Rebase and retargeting is done (I am not sure if GitHub sends an email about it).

tie commented on pkgs/development/web/nodejs/add-qemu-wrapper-support.patch:
Are there any changes from the buildroot patch? Can we use `fetchpatch2` to fetch the patch?

Cynerd replied:
No changes in the end. I will switch it to fetchpatch.

tie commented on pkgs/development/web/nodejs/nodejs.nix:
```diff
@@ -126,7 +127,8 @@ let
 
-    inherit patches;
+    patches = patches
+      ++ (lib.optional isCrossBitness ./add-qemu-wrapper-support.patch);
```
I’m not really a fan of conditional patches. Can we apply the patch unconditionally (and replace `MAYBE_WRAPPER` with an empty string if we are not using an emulator)?

Cynerd replied:
Ok.. in my eyes, it makes it a bit more complex, but ok. There is also an option to just always use a wrapper because, for the native build, it will work as well.

tie commented on pkgs/development/web/nodejs/nodejs.nix:
```diff
@@ -149,6 +151,17 @@ let
+      sed -i "s%@MAYBE_WRAPPER@%'<(PRODUCT_DIR)/v8-qemu-wrapper',%g" node.gyp
+      sed -i "s%@MAYBE_WRAPPER@%'<(PRODUCT_DIR)/v8-qemu-wrapper',%g" tools/v8_gypfiles/v8.gyp
+      mkdir -p out/Release
+      cat > out/Release/v8-qemu-wrapper << "EOF"
```
Hm, I think a symlink should be enough in this case. IIRC emulator contains arguments in addition to path iff `canExecute`.
https://github.com/NixOS/nixpkgs/blob/5d49679d00dedf21e4db29218c6ac27edad7855c/lib/systems/default.nix#L285-L295

tie replied:
Ah, OK.
```
nix-repl> (lib.systems.elaborate "x86_64-linux").canExecute (lib.systems.elaborate "i686-linux")   
true
```

Cynerd replied:
Yeah for i686 it needs to be a script.

tie replied:
FWIW I’ve opened PR #324071 to make an emulator a simple executable path in native case.

Cynerd replied:
With that it should be possible to use only just symlink 👍

Cynerd replied:
Can we proceed with this PR and update it just once #324071 is merged. It should be later only a minor change in `postPath`. What do you think?

tie commented on pkgs/development/web/nodejs/nodejs.nix:
```diff
@@ -60,8 +61,8 @@ let
-    CC_host = "cc";
-    CXX_host = "c++";
+    CC_host = if isCrossBitness then "${stdenv.cc.targetPrefix}cc" else "cc";
+    CXX_host = if isCrossBitness then "${stdenv.cc.targetPrefix}c++" else "c++";
```
Are we using `$NIX_CC` instead of `$NIX_CC_FOR_BUILD` in this case? If we don’t need C compiler for build platform, perhaps it makes sense to omit `depsBuildBuild` entirely if `isCrossBitness`?

Cynerd replied:
Yes it should be possible to just drop `depsBuildBuild` in case of `isCrossBitness`.

Would it even make sense to always use this magic for cross-compilation and not just when bitness differed? The issue with `isCrossBitness` is that as far as I can see, this is only extraction for the Nix-supported platforms, but as you linked, v8 actually supports only limited build platforms not actually matching bitness.

tie replied:
>Would it even make sense to always use this magic for cross-compilation and not just when bitness differed?

Yes, sounds reasonable to me. Always using an emulator instead of cross-compiling should make Node.js build faster, see also #321847.

Cynerd replied:
It is now converted to build always with the patch and with wrapper.

I tested it on amd64 and cross build from amd64 to armv7.

aduh95 commented:
This needs a rebase.

tie commented on pkgs/development/web/nodejs/nodejs.nix:
```diff
@@ -129,7 +127,24 @@ let
+    patches = patches ++ [
+      (fetchpatch2 {
+        name = "add-qemu-wrapper-support.patch";
+        url = "https://github.com/buildroot/buildroot/raw/ecf060a2ff93626fb5fbe1c52b8d42b6ee0fbb5b/package/nodejs/nodejs-src/0001-add-qemu-wrapper-support.patch";
```
GitHub repository is a mirror, the official Git repository is at https://gitlab.com/buildroot.org/buildroot.
```suggestion
        url = "https://gitlab.com/buildroot.org/buildroot/-/raw/ecf060a2ff93626fb5fbe1c52b8d42b6ee0fbb5b/package/nodejs/nodejs-src/0001-add-qemu-wrapper-support.patch";
```

tie commented on pkgs/development/web/nodejs/nodejs.nix:
```diff
@@ -180,14 +195,14 @@ let
       find . -path "./torque_*/**/*.o" -or -path "./v8*/**/*.o" | sort -u >files
       ${if stdenv.buildPlatform.isGnu then ''
-        ar -cqs $libv8/lib/libv8.a @files
+        $AR -cqs $libv8/lib/libv8.a @files
```
Hm, I’m not sure why this line is conditional. I thought Nixpkgs bintools wrapper handled response files, but maybe that’s not the case here for some reason? I’ll have to double-check that.

tie commented:
@Cynerd, I’ve made a few minor adjustments based on this PR and re-implemented buildroot patch in 42fa83e8d79d78df8d603048b604c2a33cca479d ([0001-nodejs-fix-cross-compilation-for-armv7l.patch](https://github.com/user-attachments/files/16240864/0001-nodejs-fix-cross-compilation-for-armv7l.patch)) so that the emulator trick is upstreamable. What do you think?

Cynerd commented:
> @Cynerd, I’ve made a few minor adjustments based on this PR and re-implemented buildroot patch in [42fa83e](https://github.com/NixOS/nixpkgs/commit/42fa83e8d79d78df8d603048b604c2a33cca479d) ([0001-nodejs-fix-cross-compilation-for-armv7l.patch](https://github.com/user-attachments/files/16240864/0001-nodejs-fix-cross-compilation-for-armv7l.patch)) so that the emulator trick is upstreamable. What do you think?

I am ok with that. I tried to compile it and ti seems to work so I would go with your solution.

tie commented:
Note for reviewers: #327653 contains commits from this PR (with slight modifications, see above) and goes further on top of these changes to use ninja for build.

@Cynerd, would it be OK to close this PR (since you’ve [approved ninja PR](https://github.com/NixOS/nixpkgs/pull/327653#pullrequestreview-2185276941))?

Cynerd commented:
@tie Yes. This can be closed.
